File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/remote/WatchmodeResolveService.kt
package com.rork.guidestreamtvandroid.data.remote

import com.rork.guidestreamtvandroid.SupabaseConfig
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * A single US streaming source for a title, from the `watchmode_resolve` edge
 * function. Mirrors iOS `WatchmodeSource`.
 */
@Serializable
data class WatchmodeSrc(
    @SerialName("source_id") val sourceId: Int = 0,
    val name: String = "",
    val type: String = "",
    val region: String? = null,
    @SerialName("web_url") val webUrl: String? = null,
    @SerialName("ios_url") val iosUrl: String? = null,
    @SerialName("tvos_url") val tvosUrl: String? = null,
    @SerialName("roku_url") val rokuUrl: String? = null,
    @SerialName("android_url") val androidUrl: String? = null,
    @SerialName("android_tv_url") val androidTvUrl: String? = null,
    val price: Double? = null,
    val format: String? = null,
    @SerialName("end_date") val endDate: String? = null,
)

/**
 * Envelope for the `watchmode_resolve` edge function response.
 */
@Serializable
data class WatchmodeResolveResponse(
    @SerialName("primary_source") val primarySource: WatchmodeSrc? = null,
    @SerialName("us_sources") val usSources: List<WatchmodeSrc> = emptyList(),
    val overview: String? = null,
    @SerialName("provider_name_fallback") val providerNameFallback: String? = null,
    @SerialName("episode_source") val episodeSource: WatchmodeSrc? = null,
    /** Region codes where the title streams, so a title with no US sources
     *  can read as unavailable instead of an empty row. */
    @SerialName("availability_regions") val availabilityRegions: List<String> = emptyList(),
)

/**
 * Client for the `watchmode_resolve` edge function (deployed with
 * verify_jwt=false). Resolves a TMDB id to its US streaming sources for the
 * title-scoped Reels switcher. Uses the same raw Ktor anon-key POST pattern as
 * `AskStreamSheet`.
 */
object WatchmodeResolveService {

    suspend fun resolve(
        tmdbId: Int,
        isTV: Boolean,
        subscribedServices: List<String> = emptyList(),
        season: Int? = null,
        episode: Int? = null,
        episodePlatformHint: String? = null,
        sourceId: Int? = null,
    ): WatchmodeResolveResponse {
        return try {
            val client = HttpClient {
                install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
            }
            try {
                val url = "${SupabaseConfig.URL.trim()}/functions/v1/watchmode_resolve"
                val decoded = post(client, url, buildBody(tmdbId, isTV, subscribedServices, season, episode, episodePlatformHint, sourceId))
                    ?: return WatchmodeResolveResponse()
                // A saved movie stored with the wrong media type resolves to
                // no sources at all — retry exactly once with isTV omitted so
                // the server probes TV then movie, and keep the retry only
                // when it found something.
                if (decoded.primarySource == null && decoded.usSources.isEmpty()) {
                    val retried = post(client, url, buildBody(tmdbId, null, subscribedServices, season, episode, episodePlatformHint, sourceId))
                    if (retried != null && (retried.primarySource != null || retried.usSources.isNotEmpty())) {
                        return retried
                    }
                }
                decoded
            } finally {
                client.close()
            }
        } catch (_: Exception) {
            WatchmodeResolveResponse()
        }
    }

    /**
     * Builds the request body, omitting null parameters rather than sending
     * nulls; a null [isTV] is omitted so the server probes both media types.
     */
    private fun buildBody(
        tmdbId: Int,
        isTV: Boolean?,
        subscribedServices: List<String>,
        season: Int?,
        episode: Int?,
        episodePlatformHint: String?,
        sourceId: Int?,
    ) = buildJsonObject {
        put("tmdbId", JsonPrimitive(tmdbId))
        if (isTV != null) put("isTV", JsonPrimitive(isTV))
        put("subscribedServices", JsonArray(subscribedServices.map { JsonPrimitive(it) }))
        if (season != null) put("season", JsonPrimitive(season))
        if (episode != null) put("episode", JsonPrimitive(episode))
        if (episodePlatformHint != null) put("episodePlatformHint", JsonPrimitive(episodePlatformHint))
        if (sourceId != null) put("sourceId", JsonPrimitive(sourceId))
    }

    /**
     * Single POST + decode. Returns null on transport failure, non-200
     * status, or decode failure so the caller never retries those.
     */
    private suspend fun post(client: HttpClient, url: String, payload: JsonObject): WatchmodeResolveResponse? {
        return try {
            val response: HttpResponse = client.post(url) {
                contentType(ContentType.Application.Json)
                header(HttpHeaders.ContentType, "application/json")
                header("apikey", SupabaseConfig.ANON_KEY)
                header(HttpHeaders.Authorization, "Bearer ${SupabaseConfig.ANON_KEY}")
                setBody(payload.toString())
            }
            if (response.status.value == 200) response.body() else null
        } catch (_: Exception) {
            null
        }
    }
}

File: ios/GuideStreamTV/Services/WatchmodeResolveService.swift
//
//  WatchmodeResolveService.swift
//  GuideStreamTV
//
//  Client for the `watchmode_resolve` Supabase edge function (deployed at
//  version 7 with verify_jwt=false, so the anon key is sufficient and no user
//  session is needed). Routes every Watchmode call through the server so the
//  hardcoded Watchmode API key never ships in the app binary — mirrors the
//  pattern already used by tvOS (TVWatchmodeResolver) and Android.
//
//  The edge function mirrors the old client-side pipeline (Watchmode id lookup,
//  title detail, TMDB provider tiebreaker, network signals, subscribed-first
//  priority, US filter, dedupe, rank) and returns the resolved primary source,
//  ranked US source list, overview, provider-name fallback, and — for episode
//  queries — a single episode source narrowed to the hinted or primary service.
//
//  Source objects use the same snake_case wire shape as Watchmode's direct
//  API, so `WatchmodeSource` is reused for decoding rather than inventing a
//  parallel model.
//

import Foundation

nonisolated enum WatchmodeResolveService {

    /// Response envelope from the `watchmode_resolve` edge function.
    struct Response: Decodable, Sendable {
        let primarySource: WatchmodeSource?
        let usSources: [WatchmodeSource]
        let overview: String?
        let providerNameFallback: String?
        let episodeSource: WatchmodeSource?
        let resolvedMediaType: String?

        enum CodingKeys: String, CodingKey {
            case primarySource = "primary_source"
            case usSources = "us_sources"
            case overview
            case providerNameFallback = "provider_name_fallback"
            case episodeSource = "episode_source"
            case resolvedMediaType = "resolved_media_type"
        }
    }

    /// NSCache wrapper so a struct value can be stored in NSCache.
    private final class ResponseBox: NSObject {
        let response: Response
        init(_ response: Response) { self.response = response }
    }

    /// In-memory cache for episode-level lookups so reopening the same
    /// episode spends no additional paid Watchmode call. The edge function
    /// has no server-side cache. Keyed by tmdbId, season, episode, and hint.
    private static let episodeCache = NSCache<NSString, ResponseBox>()

    /// Resolves streaming sources for a title via the `watchmode_resolve`
    /// edge function.
    ///
    /// - Parameters:
    ///   - tmdbId: The TMDB id of the title.
    ///   - isTV: Whether the id is a TV series. When nil the server probes
    ///     TV then movie and reports what it found in `resolvedMediaType`.
    ///   - season: Season number for episode-level lookups.
    ///   - episode: Episode number for episode-level lookups.
    ///   - episodePlatformHint: A service name to prioritise for episode
    ///     source resolution. The server narrows `episodeSource` to this
    ///     service when present.
    ///   - subscribedServices: Service names for subscribed-first priority.
    /// - Returns: The decoded `Response` on any HTTP 200 (including a
    ///   response with empty `usSources`), or `nil` only on transport
    ///   failure, non-200 status, or decode failure so callers can
    ///   distinguish a Supabase outage from an empty result.
    static func resolve(
        tmdbId: Int,
        isTV: Bool? = nil,
        season: Int? = nil,
        episode: Int? = nil,
        episodePlatformHint: String? = nil,
        sourceId: Int? = nil,
        subscribedServices: [String] = []
    ) async -> Response? {
        // Episode-level cache check — only for episode lookups (both
        // season and episode provided). sourceId is included in the key
        // so a pinned lookup never reads back a cached unpinned response.
        let isEpisodeLookup = season != nil && episode != nil
        let cacheKey = "\(tmdbId)-\(season ?? -1)-\(episode ?? -1)-\(episodePlatformHint ?? "")-\(sourceId ?? -1)" as NSString
        if isEpisodeLookup,
           let cached = episodeCache.object(forKey: cacheKey) {
            return cached.response
        }

        let base = SupabaseConfig.url.trimmingCharacters(in: .whitespaces)
        guard let url = URL(string: "\(base)/functions/v1/watchmode_resolve") else { return nil }

        // Build body, omitting nil parameters rather than sending nulls.
        var body: [String: Any] = ["tmdbId": tmdbId]
        if let isTV { body["isTV"] = isTV }
        if let season { body["season"] = season }
        if let episode { body["episode"] = episode }
        if let episodePlatformHint { body["episodePlatformHint"] = episodePlatformHint }
        if let sourceId { body["sourceId"] = sourceId }
        if !subscribedServices.isEmpty { body["subscribedServices"] = subscribedServices }

        guard let httpBody = try? JSONSerialization.data(withJSONObject: body) else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(SupabaseConfig.anonKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = httpBody

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            return nil
        }

        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
        guard let decoded = try? JSONDecoder().decode(Response.self, from: data) else { return nil }

        // Cache episode-level responses.
        if isEpisodeLookup {
            episodeCache.setObject(ResponseBox(decoded), forKey: cacheKey)
        }

        return decoded
    }
}
